### PR TITLE
Add abstract tool support

### PR DESCRIPTION
**Pull request by arjunkumargiri**
### Description
Build Abstract tool to simplify tool creation and remove boilerplate tool management implementations to abstract super class. Refactored existing tools to make use of `AbstractTool`.

Tools that extends `AbstractTool` would only need to implement `run` and `validate` methods.

### Issues Resolved
[List any issues this PR will resolve]
 
### Check List
- [x] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: spi/build.gradle [hunk 1]
```diff
@@ -7,15 +7,15 @@ import com.github.jengelman.gradle.plugins.shadow.ShadowBasePlugin
 import org.opensearch.gradle.test.RestIntegTestTask
 
 plugins {
-    id 'com.github.johnrengelman.shadow'
+    id 'java'
+    id "io.freefair.lombok"
```
Comment (Hailong-am) [inline spi/build.gradle:11]: seems that lombok is not used in spi sub project

File: spi/build.gradle [hunk 2]
```diff
@@ -7,15 +7,15 @@ import com.github.jengelman.gradle.plugins.shadow.ShadowBasePlugin
...
     id 'jacoco'
+    id 'com.github.johnrengelman.shadow'
     id 'maven-publish'
+    id 'com.diffplug.spotless' version '6.18.0'
     id 'signing'
 }
 
-apply plugin: 'opensearch.java'
-apply plugin: 'opensearch.testclusters'
-apply plugin: 'opensearch.java-rest-test'
```
Comment (Hailong-am) [inline spi/build.gradle:19]: should we remove `integTest` task at line 75 also?

File: spi/build.gradle (hunk 1 above)
Comment (arjunkumargiri) [reply spi/build.gradle]: Thanks for catching this. Done.

File: spi/build.gradle (hunk 2 above)
Comment (arjunkumargiri) [reply spi/build.gradle:18]: Done

**Comment by Hailong-am** [comment]
CI run with windows platform failed with error below
```
 What went wrong:
A problem was found with the configuration of task ':opensearch-ml-plugin:compileTestJava' (type 'JavaCompile').
  - Gradle detected a problem with the following location: 'D:\a\ml-commons\ml-commons\spi\build\libs\opensearch-ml-spi-2.11.0.0-SNAPSHOT.jar'.
> Task :opensearch-ml-search-processors:generateTestEffectiveLombokConfig
> Task :opensearch-ml-memory:generateTestEffectiveLombokConfig
    
    Reason: Task ':opensearch-ml-plugin:compileTestJava' uses this output of task ':opensearch-ml-spi:jar' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.
    
    Possible solutions:
      1. Declare task ':opensearch-ml-spi:jar' as an input of ':opensearch-ml-plugin:compileTestJava'.
      2. Declare an explicit dependency on ':opensearch-ml-spi:jar' from ':opensearch-ml-plugin:compileTestJava' using Task#dependsOn.
      3. Declare an explicit dependency on ':opensearch-ml-spi:jar' from ':opensearch-ml-plugin:compileTestJava' using Task#mustRunAfter.
    
    Please refer to https://docs.gradle.org/8.1.1/userguide/validation_problems.html#implicit_dependency for more details about this problem.

```

Try to disable jar task on spi that could help
```
shadowJar {
    archiveClassifier = null
}
jar.enabled = false
```

File: spi/src/main/java/org/opensearch/ml/common/spi/tools/AbstractTool.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
+package org.opensearch.ml.common.spi.tools;
```
Comment (ylwu-amzn) [inline spi/src/main/java/org/opensearch/ml/common/spi/tools/AbstractTool.java:1]: Add license header ?

File: spi/src/test/java/org.opensearch.ml.common.spi.tools/AbstractToolTest.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
+package org.opensearch.ml.common.spi.tools;
```
Comment (ylwu-amzn) [inline spi/src/test/java/org.opensearch.ml.common.spi.tools/AbstractToolTest.java:1]: Add license header

**Review by ylwu-amzn — APPROVED** [review]

**Comment by ylwu-amzn** [comment]
merge first to unblock others

**Pull request merged**